pgrx/src/nullable.rs

```diff
@@ -0,0 +1,523 @@
+//! Traits and implementations of functionality relating to Postgres null
+//! and nullable types, particularly in relation to arrays and other container
+//! types which implement nullable (empty slot) behavior in a
+//! "structure-of-arrays" manner
+
+use std::{fmt::Debug, iter::Enumerate, marker::PhantomData};
+
+use bitvec::slice::BitSlice;
+
+use crate::Datum;
+
+pub enum Nullable<T> {
+    Valid(T),
+    Null,
+}
+
+impl<T> Debug for Nullable<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Valid(value) => f.debug_tuple("Valid").field(value).finish(),
+            Self::Null => write!(f, "Null"),
+        }
+    }
+}
+
+impl<T> Clone for Nullable<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        match self {
+            Self::Valid(arg0) => Self::Valid(arg0.clone()),
+            Self::Null => Self::Null,
+        }
+    }
+}
+
+impl<T> PartialEq for Nullable<T>
+where
+    T: Eq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Valid(l_value), Self::Valid(r_value)) => l_value == r_value,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
+        }
+    }
+}
+
+impl<T> Eq for Nullable<T> where T: PartialEq + Eq {}
+
+impl<T> Nullable<T> {
```

Comment:
Consider `map` and `and_then`?

Reply:
Oh absolutely, these feel like a bit of a formality but I can take care of that yeah.

Reply:
Added as requested in 1a3d5719b77c233fd96440597e3aa43d794847d4